Sources/NIOExtras/HTTP1ProxyConnectHandler.swift

```diff
@@ -0,0 +1,232 @@
+        case .failed, .completed:
+            break
+        case .initialized, .connectSent, .headReceived:
+            self.state = .failed(Error.noResult)
+            self.promise?.fail(Error.noResult)
+        }
+    }
+
+    public func channelActive(context: ChannelHandlerContext) {
+        self.sendConnect(context: context)
+        context.fireChannelActive()
+    }
+
+    public func channelInactive(context: ChannelHandlerContext) {
+        switch self.state {
+        case .initialized:
+            preconditionFailure("How can we receive a channelInactive before a channelActive?")
+        case .connectSent(let timeout), .headReceived(let timeout):
+            timeout.cancel()
+            self.failWithError(Error.remoteConnectionClosed, context: context, closeConnection: false)
+
+        case .failed, .completed:
+            break
+        }
+        context.fireChannelInactive()
+    }
+
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        preconditionFailure("We don't support outgoing traffic during HTTP Proxy update.")
+    }
+
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        switch self.unwrapInboundIn(data) {
+        case .head(let head):
+            self.handleHTTPHeadReceived(head, context: context)
+        case .body:
+            self.handleHTTPBodyReceived(context: context)
+        case .end:
+            self.handleHTTPEndReceived(context: context)
+        }
+    }
+
+    private func sendConnect(context: ChannelHandlerContext) {
+        guard case .initialized = self.state else {
+            // we might run into this handler twice, once in handlerAdded and once in channelActive.
+            return
+        }
+
+        let timeout = context.eventLoop.scheduleTask(deadline: self.deadline) {
+            switch self.state {
+            case .initialized:
+                preconditionFailure("How can we have a scheduled timeout, if the connection is not even up?")
+
+            case .connectSent, .headReceived:
+                self.failWithError(Error.httpProxyHandshakeTimeout, context: context)
+
+            case .failed, .completed:
+                break
+            }
+        }
+
+        self.state = .connectSent(timeout)
+
+        let head = HTTPRequestHead(
+            version: .init(major: 1, minor: 1),
+            method: .CONNECT,
+            uri: "\(self.targetHost):\(self.targetPort)",
+            headers: self.headers
+        )
+
+        context.write(self.wrapOutboundOut(.head(head)), promise: nil)
+        context.write(self.wrapOutboundOut(.end(nil)), promise: nil)
+        context.flush()
+    }
+
+    private func handleHTTPHeadReceived(_ head: HTTPResponseHead, context: ChannelHandlerContext) {
+        guard case .connectSent(let scheduled) = self.state else {
+            preconditionFailure("HTTPDecoder should throw an error, if we have not send a request")
```

Comment:
This `preconditionFailure` is really aggressive. We should absolutely tolerate receiving the HTTP head _late_.

Reply:
Changed so that we now tolerate the state being either `.failed` or `.completed` at this point without failing a precondition.

Reply:
I think we shouldn't tolerate `.completed` either, just `.failed`. We can only be in completed after we have received a http head and shouldn't receive a second head.